src/plugins/data/common/index_patterns/utils.ts

```diff
@@ -28,26 +28,71 @@
  * under the License.
  */
 
+import { DataSourceAttributes } from 'src/plugins/data_source/common/data_sources';
 import type { IndexPatternSavedObjectAttrs } from './index_patterns';
-import type { SavedObjectsClientCommon } from '../types';
+import type { SavedObject, SavedObjectReference, SavedObjectsClientCommon } from '../types';
 
 /**
  * Returns an object matching a given title
  *
  * @param client {SavedObjectsClientCommon}
  * @param title {string}
+ * @param dataSourceId {string}{optional}
  * @returns {Promise<SavedObject|undefined>}
  */
-export async function findByTitle(client: SavedObjectsClientCommon, title: string) {
+export async function findByTitle(
+  client: SavedObjectsClientCommon,
+  title: string,
+  dataSourceId?: string
+) {
   if (title) {
-    const savedObjects = await client.find<IndexPatternSavedObjectAttrs>({
-      type: 'index-pattern',
-      perPage: 10,
-      search: `"${title}"`,
-      searchFields: ['title'],
-      fields: ['title'],
+    const savedObjects = (
+      await client.find<IndexPatternSavedObjectAttrs>({
+        type: 'index-pattern',
+        perPage: 10,
+        search: `"${title}"`,
+        searchFields: ['title'],
+        fields: ['title'],
+      })
+    ).filter((obj) => {
+      return obj && obj.attributes && validateDataSourceReference(obj, dataSourceId);
     });
 
     return savedObjects.find((obj) => obj.attributes.title.toLowerCase() === title.toLowerCase());
   }
 }
+
+// This is used to validate datasource reference of index pattern
+export const validateDataSourceReference = (
```

Comment:
This util function is to filter the index pattern results that comes with the datasource reference we want. It's currently looking for the first element in the array, cuz I am not aware of any other cases that there could be other type of reference in an index-pattern object, or more than 1 element in the references array. Do you know any of that?

But you are raising a valid point, since references is just json fields that doesn't have strict mapping. I can modify the condition check in this utils function like below.

```
  const references = indexPattern.references;
  if (dataSourceId) {
    return references.some((ref) => ref.id === dataSourceId && ref.type === 'data-source');
  } else {
```

Reply:
I feel we could just get the` data-source` from the reference? If we find one then return, if not then it's a default cluster. Basically we don't do the validation to enforce the content of the index pattern reference. What do you think?

Reply:
datasource id must match in our case. Otherwise, it could return multiple results and fail to let us create index pattern with duplicate name.